### PR TITLE
Update Ruby to 3.2

### PR DESCRIPTION
**Pull request by ashie**
Since Ruby 2.7 is close to EOL, we have to update Ruby.
The next major version of td-agent will bundle Ruby 3.2.

* Bundle Ruby 3.2 and remove older Ruby sources
* Add libyaml-devel and libffi-devel package since Ruby 3.2 no longer bundles these sources
* Add Rust to build YJIT of Ruby
* Remove obsolete distributions from default build targets
* Update nokogiri to 1.14.0 to make it buildable on Windows
* Bump up the version number to 5.0.0 to indicate that the master branch is going to develop next major version

FYI: https://www.ruby-lang.org/en/news/2022/12/25/ruby-3-2-0-released/

TODO:
  * Add missing libs for macOS (but I won't address it in this PR, I'll track it at #421)
  * ~~Fix CI failure on Windows~~ Done

### CODE REVIEW

**Comment by ashie** [comment]
Hmm, `bundle install` stalls after completing to install all gems on Ruby 3.2 on Windows:
```
...
Fetching fluent-plugin-rewrite-tag-filter 2.4.0
Fetching fluent-plugin-calyptia-monitoring 0.1.3
Installing fluent-plugin-rewrite-tag-filter 2.4.0
Installing fluent-plugin-calyptia-monitoring 0.1.3
Fetching fluent-plugin-prometheus_pushgateway 0.1.0
Installing fluent-plugin-prometheus_pushgateway 0.1.0
(no output then timed out)
```
Updating bundler to 2.4.4 doesn't take effect.

It doesn't reproduce on Ruby 3.1:
```
...
Installing fluent-plugin-prometheus_pushgateway 0.1.0
Bundle complete! 63 Gemfile dependencies, 110 gems now installed.
...
```

**Comment by ashie** [comment]
Run `rake msi:selfbuild` and break by `C-c` after stall:
```
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/worker.rb:108:in `block in add_interrupt_handler'
<internal:thread_sync>:18:in `pop'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/worker.rb:42:in `deq'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:173:in `process_specs'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:133:in `install_with_worker'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:89:in `call'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:67:in `call'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:244:in `install_in_parallel'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:201:in `install'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:89:in `block in run'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/process_lock.rb:12:in `block in lock'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/process_lock.rb:9:in `open'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/process_lock.rb:9:in `lock'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:71:in `run'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:23:in `install'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/cli/install.rb:62:in `run'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:260:in `block in install'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/settings.rb:131:in `temporary'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:259:in `install'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:34:in `dispatch'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:28:in `start'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/exe/bundle:45:in `block in <top (required)>'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
C:/Users/aho/Projects/td-agent-builder/td-agent/staging/lib/ruby/gems/3.2.0/gems/bundler-2.4.4/exe/bundle:33:in `<top (required)>'
```

**Comment by ashie** [comment]
When I run `bundler install` with `--jobs=1`, it stalls on installing nokogiri:
```
Fetching nokogiri 1.13.9
Installing nokogiri 1.13.9 with native extensions
```
(It reproduces on `msi:build` but doesn't reproduce on `msi:selfbuild`.)

It might be a same issue we experienced in the past.

**Comment by ashie** [comment]
I've confirmed on my local environment that nokogiri 1.14.0 is buildable on Ruby 3.2 for Windows.

**Comment by cosmo0920** [comment]
> It might be a same issue we experienced in the past.

This could be occurred with fat gem mechanism. Maybe, nokogiri 1.13 or older does not include a binary for Ruby3.2, right?

**Comment by ashie** [comment]
> This could be occurred with fat gem mechanism. Maybe, nokogiri 1.13 or older does not include a binary for Ruby3.2, right?

Yes. Now I succeeded to build it also on CI by nokogiri 1.14.0.
The next problem is that installation test is failed on CI.
But the built package is installable on my & @daipom's local environment.
Probably it fails to launch the fluentwinsvc service on CI.
I'm suspecting that required runtime library is lacked in CI env (or implicitly installed on our environment?).

**Comment by cosmo0920** [comment]
> I'm suspecting that required runtime library is lacked in CI env (it's implicitly installed on our environment?).

I recommend to use testing msi package or zip package on Windows Sandbox. This is provided from Windows Professional Edition or later. However, this isolated and refreshed in every time environment quite useful to test implicit dependencies. Or, using Dependencies to testify dependencies of Windows executables: https://github.com/lucasg/Dependencies

**Comment by cosmo0920** [comment]
This is because YJIT will introduce MSVC runtime dlls.
Ruby's YJIT does not link CRT as static.

https://github.com/ruby/ruby/blob/master/yjit/Cargo.toml

**Comment by ashie** [comment]
> I recommend to use Windows Sandbox for testing msi package or zip package. 

Thanks for your information. I forgot it so far.
But even I tried on Windows Sandbox, it succeeded to install...

**Comment by ashie** [comment]
> This is because YJIT will introduce MSVC runtime dlls. Ruby's YJIT does not link CRT as static.

In my understanding, Ruby on Windows doesn't support YJIT.
https://www.ruby-lang.org/ja/news/2022/12/25/ruby-3-2-0-released/

In actual, RubyInstaller 3.2 doesn't provide `--yjit` option.
```
C:\opt\td-agent>ruby --version
ruby 3.2.0 (2022-12-25 revision a528908271) [x64-mingw-ucrt]

C:\opt\td-agent>ruby --yjit
ruby: warning: Ruby was built without YJIT support. You may need to install rustc to build Ruby with YJIT.
```

**Comment by cosmo0920** [comment]
Got it. Interesting...

**Comment by ashie** [comment]
I completely misunderstood the error, it's not installation error:

```
  1) Package "td-agent v5.0.0" is expected to be installed
     Failure/Error: it { should be_installed }
     NoMethodError:
       undefined method `exists?' for File:Class

     # ./serverspec/windows/td-agent.rb:12:in `block (2 levels) in <top (required)>'
```

Something uses `File.exists?` which is removed as of Ruby 3.2:
https://www.ruby-lang.org/ja/news/2022/12/25/ruby-3-2-0-released/

One of suspected module is serverspec but it's already fixed in 2.42.1:
https://github.com/mizzy/serverspec/commit/d75dbb46203a1b2f99767a321e17d02e455a4ca4

**Comment by daipom** [comment]
this? https://github.com/fluent/fluent-package-builder/blob/master/td-agent/Rakefile#L87

**Comment by daipom** [comment]
Oh, sorry, this is not related to serverspec test...

**Comment by ashie** [comment]
I got it. It's specinfra's issue:
```
  48) gem files Package "ffi" is expected to be installed by "gem" with version "1.15.5"
      Failure/Error: it { should be_installed.by("gem").with_version(gem.version) }
      NoMethodError:
        undefined method `exists?' for File:Class

      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:73:in `block in find_powershell'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:73:in `each'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:73:in `find'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:73:in `find_powershell'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:65:in `x64_powershell'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:59:in `powershell'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/backend/cmd.rb:14:in `run_command'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/runner.rb:27:in `run'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/specinfra-2.84.0/lib/specinfra/runner.rb:19:in `method_missing'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/serverspec-2.42.1/lib/serverspec/type/package.rb:9:in `installed?'
      # C:/Users/aho/.gem/ruby/3.2.0/gems/serverspec-2.42.1/lib/serverspec/matcher/be_installed.rb:6:in `block (2 levels) in <top (required)>'
      # ./serverspec/windows/td-agent.rb:26:in `block (4 levels) in <top (required)>'
```

**Comment by ashie** [comment]
Fixed in specinfra-2.84.1

**Comment by daipom** [comment]
Wow! Super quick!

File: td-agent/msi/Dockerfile [hunk 1]
```diff
@@ -26,8 +26,8 @@ RUN @"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -I
 RUN \
   choco install -y git wixtoolset 7zip & \
   choco install -y cmake --installargs 'ADD_CMAKE_TO_PATH=System' && \
-  choco install -y msys2 --params /NoUpdate --version=20220904.0.0 && \
-  choco install ruby -y --version=3.0.4.1 && \
+  choco install -y msys2 --params /NoUpdate --version=20221216.0.0 && \
```
Comment (cosmo0920) [inline td-agent/msi/Dockerfile:29]: The latest msys2 version is 20230127.0.0. Shouldn't we use this or staying the previous one?

**Review by cosmo0920 — COMMENTED** [review]
Almost of all really good to me. However, the latest msys2 is not same for currently used in this PR. Could you confirm whether we can use the latest version of msys2 with this PR or not?

**Comment by daipom** [comment]
Sorry, please let me see this tomorrow.

**Comment by kenhys** [comment]
If it doesn't cause a regression, adopting latest openssl 1.1.1s may be better.
https://mta.openssl.org/pipermail/openssl-announce/2022-November/000242.html

NOTE: openssl 1.1.1r has a regression. so announced of withdrawal
https://mta.openssl.org/pipermail/openssl-announce/2022-October/000237.html

File: td-agent/msi/Dockerfile (hunk 1 above)
Comment (kenhys) [reply td-agent/msi/Dockerfile:29]: With 20230127.0.0, it seems that .msi could be built successfully.

**Review by kenhys — APPROVED** [review]
LGTM.

File: td-agent/config.rb [hunk 2]
```diff
@@ -13,42 +13,25 @@
 # https://www.openssl.org/source/
 OPENSSL_VERSION = "1.1.1q"
```
[the remark this answers is not in the document]
Comment (ashie) [reply td-agent/config.rb:14]: This PR is focused on updating Ruby.
I'll update it in another PR.

**Comment by ashie** [comment]
> If it doesn't cause a regression, adopting latest openssl 1.1.1s may be better. https://mta.openssl.org/pipermail/openssl-announce/2022-November/000242.html
> 
> NOTE: openssl 1.1.1r has a regression. so announced of withdrawal https://mta.openssl.org/pipermail/openssl-announce/2022-October/000237.html

This PR is focused on updating Ruby.
So I'll update OpenSSL for macOS in another PR.
For windows, OpenSSL 3.0 bundled with RubyInstaller 3.2 will be used.

File: td-agent/msi/Dockerfile (hunk 1 above)
Comment (ashie) [reply td-agent/msi/Dockerfile]: Thanks, I didn't notice it.
I've updated it.

**Review by cosmo0920 — APPROVED** [review]

**Review by daipom — APPROVED** [review]
Sorry for being late.
I have checked the fixes.
In addition, I have checked that the `--yjit` option of the embedded Ruby works in my M1 MacBook.

LGTM.

**Pull request merged**